src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1263,24 +1263,34 @@ DetachPartitionCommandList(void)
 
 /*
  * SyncMetadataToNodes tries recreating the metadata snapshot in the
- * metadata workers that are out of sync. Returns false if synchronization
- * to at least one of the workers fails.
+ * metadata workers that are out of sync. Returns the result of
+ * synchronization.
  */
-bool
+MetadataSyncResult
 SyncMetadataToNodes(void)
 {
 	List *workerList = NIL;
 	ListCell *workerCell = NULL;
-	bool result = true;
+	MetadataSyncResult result = METADATA_SYNC_SUCCESS;
 
 	if (!IsCoordinator())
 	{
-		return true;
+		return METADATA_SYNC_SUCCESS;
 	}
 
-	LockRelationOid(DistNodeRelationId(), ExclusiveLock);
+	/*
+	 * We request a ShareLock here, because we want to avoid running at
+	 * the same time master_update_node() or some other function is updating
+	 * pg_dist_node, but we should be able to run concurrently with other
+	 * functions that are scanning pg_dist_node.
+	 */
+	if (!ConditionalLockRelationOid(DistNodeRelationId(), ShareLock))
```

Comment:
maybe `RowExclusiveLock`? (which is a lot lighter than ShareLock, but still blocked by ExclusiveLock)

Reply:
or even AccessShareLock just to make sure it never blocks.

Reply:
Yeah, I was mixing up `ShareLock` with `RowShareLock` in discussion

Reply:
Looking into it more, I think we should first do some basic check using an AccessShareLock to see if there is any work to do and only take more aggressive locks if there is. `master_add_node` takes an exclusive lock and can take a really long time if it's copying reference tables to the new node and we don't want deadlock detection to be blocked during that period.

Reply:
done